File: api/rust/examples/example_config.rs
// You should glob import these to prevent your config from getting cluttered.
use pinnacle_api::prelude::*;
use pinnacle_api::*;

fn main() {
    // Connect to the Pinnacle server.
    // This needs to be called before you start calling any config functions.
    pinnacle_api::connect().unwrap();

    let mod_key = Modifier::Ctrl;

    let terminal = "alacritty";

    process::set_env("MOZ_ENABLE_WAYLAND", "1");

    // You must create a callback_vec to hold your callbacks.
    // Rust is not Lua, so it takes a bit more work to get captures working.
    //
    // Anything that requires a callback will also require a mut reference to this struct.
    //
    // Additionally, all callbacks also take in `&mut CallbackVec`.
    // This allows you to call functions that need callbacks within other callbacks.
    let mut callback_vec = CallbackVec::new();

    // Keybinds.

    input::mousebind(
        &[mod_key],
        MouseButton::Left,
        MouseEdge::Press,
        move |_| {
            window::begin_move(MouseButton::Left);
        },
        &mut callback_vec,
    );

    input::mousebind(
        &[mod_key],
        MouseButton::Right,
        MouseEdge::Press,
        move |_| {
            window::begin_resize(MouseButton::Right);
        },
        &mut callback_vec,
    );

    input::keybind(
        &[mod_key, Modifier::Alt],
        'q',
        |_| pinnacle::quit(),
        &mut callback_vec,
    );

    input::keybind(
        &[mod_key, Modifier::Alt],
        'c',
        move |_| {
            if let Some(window) = window::get_focused() {
                window.close();
            }
        },
        &mut callback_vec,
    );

    input::keybind(
        &[mod_key],
        xkbcommon::xkb::keysyms::KEY_Return,
        move |_| {
            process::spawn(vec![terminal]).unwrap();
        },
        &mut callback_vec,
    );

    input::keybind(
        &[mod_key, Modifier::Alt],
        xkbcommon::xkb::keysyms::KEY_space,
        move |_| {
            if let Some(window) = window::get_focused() {
                window.toggle_floating();
            }
        },
        &mut callback_vec,
    );

    input::keybind(
        &[mod_key],
        'f',
        move |_| {
            if let Some(window) = window::get_focused() {
                window.toggle_fullscreen();
            }
        },
        &mut callback_vec,
    );

    input::keybind(
        &[mod_key],
        'm',
        move |_| {
            if let Some(window) = window::get_focused() {
                window.toggle_maximized();
            }
        },
        &mut callback_vec,
    );

    let tags = ["1", "2", "3", "4", "5"];

    output::connect_for_all(
        move |output, _| {
            tag::add(&output, tags.as_slice());
            tag::get("1", Some(&output)).unwrap().toggle();
        },
        &mut callback_vec,
    );

    let mut layout_cycler = tag::layout_cycler(&[
        Layout::MasterStack,
        Layout::Dwindle,
        Layout::Spiral,
        Layout::CornerTopLeft,
        Layout::CornerTopRight,
        Layout::CornerBottomLeft,
        Layout::CornerBottomRight,
    ]);

    input::keybind(
        &[mod_key],
        xkbcommon::xkb::keysyms::KEY_space,
        move |_| {
            (layout_cycler.next)(None);
        },
        &mut callback_vec,
    );

    input::keybind(
        &[mod_key, Modifier::Shift],
        xkbcommon::xkb::keysyms::KEY_space,
        move |_| {
            (layout_cycler.prev)(None);
        },
        &mut callback_vec,
    );

    // Keybinds for tags

    for tag_name in tags.iter().map(|t| t.to_string()) {
        let t = tag_name.clone();
        input::keybind(
            &[mod_key],
            tag_name.chars().next().unwrap(),
            move |_| {
                tag::get(&t, None).unwrap().switch_to();
            },
            &mut callback_vec,
        );
        let t = tag_name.clone();
        input::keybind(
            &[mod_key, Modifier::Shift],
            tag_name.chars().next().unwrap(),
            move |_| {
                tag::get(&t, None).unwrap().toggle();
            },
            &mut callback_vec,
        );
        let t = tag_name.clone();
        input::keybind(
            &[mod_key, Modifier::Alt],
            tag_name.chars().next().unwrap(),
            move |_| {
                if let Some(window) = window::get_focused() {
                    window.move_to_tag(&tag::get(&t, None).unwrap());
                }
            },
            &mut callback_vec,
        );
        let t = tag_name.clone();
        input::keybind(
            &[mod_key, Modifier::Shift, Modifier::Alt],
            tag_name.chars().next().unwrap(),
            move |_| {
                if let Some(window) = window::get_focused() {
                    window.toggle_tag(&tag::get(&t, None).unwrap());
                }
            },
            &mut callback_vec,
        );
    }

    // At the very end of your config, you will need to start listening to Pinnacle in order for
    // your callbacks to be correctly called.
    //
    // This will not return unless an error occurs.
    pinnacle_api::listen(callback_vec);
}

File: api/lua_grpc/examples/default/example_config.lua
require("pinnacle").setup(function(Pinnacle)
    local Input = Pinnacle.input
    local Process = Pinnacle.process
    local Output = Pinnacle.output
    local Tag = Pinnacle.tag
    local Window = Pinnacle.window

    local key = Input.key

    ---@type Modifier
    local mod_key = "ctrl"

    local terminal = "alacritty"

    Input:set_mousebind({ mod_key }, "left", "press", function()
        Window:begin_move("left")
    end)

    Input:set_mousebind({ mod_key }, "right", "press", function()
        Window:begin_resize("right")
    end)

    ------

    Input:set_keybind({ mod_key, "alt" }, "q", function()
        print("GOT QUIT")
        Pinnacle:quit()
    end)

    Input:set_keybind({ mod_key, "alt" }, "c", function()
        local focused = Window:get_focused()
        if focused then
            focused:close()
        end
    end)

    Input:set_keybind({ mod_key }, key.Return, function()
        Process:spawn(terminal)
    end)

    Input:set_keybind({ mod_key, "alt" }, key.space, function()
        local focused = Window:get_focused()
        if focused then
            focused:toggle_floating()
        end
    end)

    Input:set_keybind({ mod_key }, "f", function()
        local focused = Window:get_focused()
        if focused then
            focused:toggle_fullscreen()
        end
    end)

    Input:set_keybind({ mod_key }, "m", function()
        local focused = Window:get_focused()
        if focused then
            focused:toggle_maximized()
        end
    end)

    local tag_names = { "1", "2", "3", "4", "5" }

    Output:connect_for_all(function(op)
        local tags = Tag:add(op, tag_names)
        tags[1]:set_active(true)
    end)

    Process:spawn_once("foot")

    local layout_cycler = Tag:new_layout_cycler({
        "master_stack",
        "dwindle",
        "spiral",
        "corner_top_left",
        "corner_top_right",
        "corner_bottom_left",
        "corner_bottom_right",
    })

    Input:set_keybind({ mod_key }, key.space, function()
        local focused_op = Output:get_focused()
        if focused_op then
            layout_cycler.next(focused_op)
        end
    end)

    Input:set_keybind({ mod_key, "shift" }, key.space, function()
        local focused_op = Output:get_focused()
        if focused_op then
            layout_cycler.prev(focused_op)
        end
    end)

    for _, tag_name in ipairs(tag_names) do
        -- nil-safety: tags are guaranteed to be on the outputs due to connect_for_all above
        Input:set_keybind({ mod_key }, tag_name, function()
            Tag:get(tag_name):switch_to()
        end)
        Input:set_keybind({ mod_key, "shift" }, tag_name, function()
            Tag:get(tag_name):toggle_active()
        end)
        Input:set_keybind({ mod_key, "alt" }, tag_name, function()
            local focused = Window:get_focused()
            if focused then
                focused:move_to_tag(Tag:get(tag_name) --[[@as TagHandle]])
            end
        end)
        Input:set_keybind({ mod_key, "shift", "alt" }, tag_name, function()
            local focused = Window:get_focused()
            if focused then
                focused:toggle_tag(Tag:get(tag_name) --[[@as TagHandle]])
            end
        end)
    end
end)
